extra_data/components.py

```diff
@@ -170,18 +170,19 @@ def __repr__(self):
         )
 
     @staticmethod
-    def _fill_value(value, dtype):
-        if value is None:
-            if dtype.kind != 'f':
-                value = 0
-            else:
-                value = np.nan
+    def _concat(arrays, index, fill_value, astype):
+        dtype = arrays[0].dtype if astype is None else np.dtype(astype)
+        if fill_value is None:
+            fill_value = np.nan if dtype.kind == 'f' else 0
+        fill_value = dtype.type(fill_value)
 
-        # enforce that fill value is compatible with array dtype
-        value = dtype.type(value)
-        return value
+        return xarray.concat(
+            arrays,
+            pd.Index(index, name='module'),
+            fill_value=fill_value
+        ).astype(dtype, copy=False)
```

Comment:
```suggestion
        return xarray.concat(
            [a.astype(dtype, copy=False) for a in arrays],
            pd.Index(index, name='module'),
            fill_value=fill_value
        )
```

It's not actually documented (or not obviously) that `xarray.concat` will convert the dtype if given `fill_value=nan`. I guess it's unlikely to change, but doing astype before that ensures it would still do the right thing if xarray ever changed that.